README.md
# <img src="./sticker.svg" width="30%" align="right" /> BioTools

 [![Latest Release](https://img.shields.io/github/release/BioJulia/BioTools.jl.svg?style=flat-square)](https://github.com/BioJulia/BioTools.jl/releases/latest)
 [![MIT license](https://img.shields.io/badge/license-MIT-green.svg?style=flat-square)](https://github.com/BioJulia/BioTools.jl/blob/master/LICENSE)
 [![Stable documentation](https://img.shields.io/badge/docs-stable-blue.svg?style=flat-square)](https://biojulia.github.io/BioTools.jl/stable)
 [![Latest documentation](https://img.shields.io/badge/docs-latest-blue.svg?style=flat-square)](https://biojulia.github.io/BioTools.jl/latest)
![Lifecycle](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg?style=flat-square)
[![Chat on Discord](https://img.shields.io/badge/discord-chat-blue.svg?style=flat-square&logo=discord&colorB=%237289DA)](https://discord.gg/z73YNFz)


## Description

BioTools provides interfaces to common external biological tools from julia scripts
and programs.

## Installation

Install BioTools from the Julia REPL:

```julia
using Pkg
add("BioTools")
# Pkg.add("BioTools") for julia v0.6-
```

If you are interested in the cutting edge of the development, please check out
the master branch to try new features before release.

## Testing

BioTools is tested against Julia `0.7` and current `1.X` on Linux, OS X, and Windows.

| **PackageEvaluator**                                            | **Latest Build Status**                                                                                |
|:---------------------------------------------------------------:|:------------------------------------------------------------------------------------------------------:|
| [![](https://pkg.julialang.org/badges/BioTools_0.6.svg)](https://pkg.julialang.org/detail/BioTools) [![](https://pkg.julialang.org/badges/BioTools_0.7.svg)](https://pkg.julialang.org/detail/BioTools) | [![](https://img.shields.io/travis/BioJulia/BioTools.jl/master.svg?label=Linux+/+macOS)](https://travis-ci.org/BioJulia/BioTools.jl) [![](https://ci.appveyor.com/api/projects/status/j9ikh4s0914ke29b?svg=true)](https://ci.appveyor.com/project/Ward9250/biotools-jl/branch/master) [![](https://codecov.io/gh/BioJulia/BioTools.jl/branch/master/graph/badge.svg)](https://codecov.io/gh/BioJulia/BioTools.jl) |

## Contributing

We appreciate contributions from users including reporting bugs, fixing
issues, improving performance and adding new features.

Take a look at the [CONTRIBUTING](CONTRIBUTING.md) file provided with
every BioJulia package package for detailed contributor and maintainer
guidelines.


### Financial contributions

We also welcome financial contributions in full transparency on our
[open collective](https://opencollective.com/biojulia).
Anyone can file an expense. If the expense makes sense for the development
of the community, it will be "merged" in the ledger of our open collective by
the core contributors and the person who filed the expense will be reimbursed.


## Backers & Sponsors

Thank you to all our backers and sponsors!

Love our work and community? [Become a backer](https://opencollective.com/biojulia#backer).

[![backers](https://opencollective.com/biojulia/backers.svg?width=890)](https://opencollective.com/biojulia#backers)

Does your company use BioJulia? Help keep BioJulia feature rich and healthy by
[sponsoring the project](https://opencollective.com/biojulia#sponsor)
Your logo will show up here with a link to your website.

[![](https://opencollective.com/biojulia/sponsor/0/avatar.svg)](https://opencollective.com/biojulia/sponsor/0/website)
[![](https://opencollective.com/biojulia/sponsor/1/avatar.svg)](https://opencollective.com/biojulia/sponsor/1/website)
[![](https://opencollective.com/biojulia/sponsor/2/avatar.svg)](https://opencollective.com/biojulia/sponsor/2/website)
[![](https://opencollective.com/biojulia/sponsor/3/avatar.svg)](https://opencollective.com/biojulia/sponsor/3/website)
[![](https://opencollective.com/biojulia/sponsor/4/avatar.svg)](https://opencollective.com/biojulia/sponsor/4/website)
[![](https://opencollective.com/biojulia/sponsor/5/avatar.svg)](https://opencollective.com/biojulia/sponsor/5/website)
[![](https://opencollective.com/biojulia/sponsor/6/avatar.svg)](https://opencollective.com/biojulia/sponsor/6/website)
[![](https://opencollective.com/biojulia/sponsor/7/avatar.svg)](https://opencollective.com/biojulia/sponsor/7/website)
[![](https://opencollective.com/biojulia/sponsor/8/avatar.svg)](https://opencollective.com/biojulia/sponsor/8/website)
[![](https://opencollective.com/biojulia/sponsor/9/avatar.svg)](https://opencollective.com/biojulia/sponsor/9/website)


## Questions?

If you have a question about contributing or using BioJulia software, come
on over and chat to us on [Discord](https://discord.gg/z73YNFz), or you can try the
[Bio category of the Julia discourse site](https://discourse.julialang.org/c/domain/bio).

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v1.1.0] - 2022-02-12
- Allow `blastn` and `blastp` to run a `Vector{BioSequence}` against a file or database
- General maintenance.

## [v1.0.0] - 2018-12-11
- Adds support to Julia 1.0

## [v0.1.0] - 2018-02-20
**Merged pull requests:**

- Readme and docs [\#2](https://github.com/BioJulia/BioTools.jl/pull/2) ([Ward9250](https://github.com/Ward9250))
- Source code and tests [\#1](https://github.com/BioJulia/BioTools.jl/pull/1) ([Ward9250](https://github.com/Ward9250))

[Unreleased]: https://github.com/BioJulia/BioTools.jl/compare/v1.1.0...HEAD
[1.1.0]: https://github.com/BioJulia/BioTools.jl/compare/v1.0.0...v1.1.0
[1.0.0]: https://github.com/BioJulia/BioTools.jl/compare/v0.1.0...v1.0.0
[0.1.0]: https://github.com/BioJulia/BioTools.jl/tree/v0.1.0


test/runtests.jl
module TestTools

using Test

using BioSequences,
      BioTools.BLAST

import BioCore.Testing:
    get_bio_fmt_specimens

fmtdir = get_bio_fmt_specimens()

if !Sys.iswindows()  # temporarily disable the BLAST tests on Windows (issue: #197)

@testset "BLAST+ blastn" begin
    na1 = dna"""
    CGGACCAGACGGACACAGGGAGAAGCTAGTTTCTTTCATGTGATTGANAT
    NATGACTCTACTCCTAAAAGGGAAAAANCAATATCCTTGTTTACAGAAGA
    GAAACAAACAAGCCCCACTCAGCTCAGTCACAGGAGAGAN
    """
    na2 = dna"""
    CGGAGCCAGCGAGCATATGCTGCATGAGGACCTTTCTATCTTACATTATG
    GCTGGGAATCTTACTCTTTCATCTGATACCTTGTTCAGATTTCAAAATAG
    TTGTAGCCTTATCCTGGTTTTACAGATGTGAAACTTTCAA
    """
    fna = joinpath(fmtdir, "FASTA", "f002.fasta")
    nucldb = joinpath(fmtdir, "BLASTDB", "f002")
    nuclresults = joinpath(fmtdir, "BLASTDB", "f002.xml")

    @test typeof(blastn(na1, na2)) == Array{BLASTResult, 1}
    @test typeof(blastn(na1, [na1, na2])) == Array{BLASTResult, 1}
    @test typeof(blastn([na1, na2], [na1, na2])) == Array{BLASTResult, 1}
    @test typeof(blastn(na1, nucldb, db=true)) == Array{BLASTResult, 1}
    @test typeof(blastn(na1, fna)) == Array{BLASTResult, 1}
    @test typeof(blastn(fna, nucldb, db=true)) == Array{BLASTResult, 1}
    @test typeof(blastn([na1, na2], nucldb, db=true)) == Array{BLASTResult, 1}
    @test typeof(blastn([na1, na2], fna)) == Array{BLASTResult, 1}
    @test typeof(blastn(fna, [na1, na2])) == Array{BLASTResult, 1}
end

@testset "BLAST+ blastp" begin
    aa1 = aa"""
    MWATLPLLCAGAWLLGVPVCGAAELSVNSLEKFHFKSWMSKHRKTYSTEE
    YHHRLQTFASNWRKINAHNNGNHTFKMALNQFSDMSFAEIKHKYLWSEPQ
    NCSATKSNYLRGTGPYPPSVDWRKKGNFVSPVKNQGACGS
    """
    aa2 = aa"""
    MWTALPLLCAGAWLLSAGATAELTVNAIEKFHFTSWMKQHQKTYSSREYS
    HRLQVFANNWRKIQAHNQRNHTFKMGLNQFSDMSFAEIKHKYLWSEPQNC
    SATKSNYLRGTGPYPSSMDWRKKGNVVSPVKNQGACGSCW
    """
    faa = joinpath(fmtdir, "FASTA", "cysprot.fasta")
    protdb = joinpath(fmtdir, "BLASTDB", "cysprot")
    protresults = joinpath(fmtdir, "BLASTDB", "cysprot.xml")

    @test typeof(blastp(aa1, aa2)) == Array{BLASTResult, 1}
    @test typeof(blastp(aa1, [aa1, aa2])) == Array{BLASTResult, 1}
    @test typeof(blastp([aa1, aa2], [aa1, aa2])) == Array{BLASTResult, 1}
    @test typeof(blastp(aa1, protdb, db=true)) == Array{BLASTResult, 1}
    @test typeof(blastp(aa1, faa)) == Array{BLASTResult, 1}
    @test typeof(blastp(faa, protdb, db=true)) == Array{BLASTResult, 1}
    @test typeof(blastp([aa1, aa2], protdb, db=true)) == Array{BLASTResult, 1}
    @test typeof(blastp([aa1, aa2], faa)) == Array{BLASTResult, 1}
    @test typeof(blastp(faa, [aa1, aa2])) == Array{BLASTResult, 1}
end

end  # if !is_windows()

end # TestTools


src/blast/blastcommandline.jl
# BLAST+ Wrapper
# ==============
#
# Wrapper for BLAST+ command line functions.
#
# This file is a part of BioJulia.
# License is MIT: https://github.com/BioJulia/Bio.jl/blob/master/LICENSE.md

struct BLASTResult
    bitscore::Float64
    expect::Float64
    queryname::String
    hitname::String
    hit::BioSequence
    alignment::AlignedSequence
end

"""
    readblastXML(blastrun::AbstractString)
Parse XML output of a blast run. Input is an XML string eg:
```julia
results = read(open("blast_results.xml"), String)
readblastXML(results)
```
Returns Vector{BLASTResult} with the sequence of the hit, the Alignment with query sequence, bitscore and expect value
"""
function readblastXML(blastrun::AbstractString; seqtype="nucl")
    dc = EzXML.parsexml(blastrun)
    rt = EzXML.root(dc)
    results = BLASTResult[]
    for iteration in findall("/BlastOutput/BlastOutput_iterations/Iteration", rt)
        queryname = EzXML.nodecontent(findfirst("Iteration_query-def", iteration))
        for hit in findall("Iteration_hits", iteration)
            if EzXML.countelements(hit) > 0
                hitname = EzXML.nodecontent(findfirst("./Hit/Hit_def", hit))
                hsps = findfirst("./Hit/Hit_hsps", hit)
                if seqtype == "nucl"
                    qseq = DNASequence(EzXML.nodecontent(findfirst("./Hsp/Hsp_qseq", hsps)))
                    hseq = DNASequence(EzXML.nodecontent(findfirst("./Hsp/Hsp_hseq", hsps)))
                elseif seqtype == "prot"
                    qseq = AminoAcidSequence(EzXML.nodecontent(findfirst("./Hsp/Hsp_qseq", hsps)))
                    hseq = AminoAcidSequence(EzXML.nodecontent(findfirst("./Hsp/Hsp_hseq", hsps)))
                else
                    throw(error("Please use \"nucl\" or \"prot\" for seqtype"))
                end

                aln = AlignedSequence(qseq, hseq)
                bitscore = parse(Float64, EzXML.nodecontent(findfirst("./Hsp/Hsp_bit-score", hsps)))
                expect = parse(Float64, EzXML.nodecontent(findfirst("./Hsp/Hsp_evalue", hsps)))
                push!(results, BLASTResult(bitscore, expect, queryname, hitname, hseq, aln))
            end
        end
    end
    return results
end

"""
`readblastXML(blastrun::Cmd)`
Parse command line blast query with XML output. Input is the blast command line command, eg:
```julia
blastresults = `blastn -query seq1.fasta -db some_database -outfmt 5`
readblastXML(blastresults)
```
Returns Vector{BLASTResult} with the sequence of the hit, the Alignment with query sequence, bitscore and expect value
"""
function readblastXML(blastrun::Cmd; seqtype="nucl")
    return readblastXML(read(blastrun, String), seqtype=seqtype)
end


"""
`blastn(query, subject, flags...)``
Runs blastn on `query` against `subject`.
    Subjects and queries may be file names (as strings), DNASequence type or
    Array of DNASequence.
    May include optional `flag`s such as `["-perc_identity", 95,]`. Do not use `-outfmt`.
"""
function blastn(query::AbstractString, subject::AbstractString, flags=[]; db::Bool=false)
    if db
        results = readblastXML(`blastn -query $query -db $subject $flags -outfmt 5`)
    else
        results = readblastXML(`blastn -query $query -subject $subject $flags -outfmt 5`)
    end
    return results
end

function blastn(query::DNASequence, subject::DNASequence, flags=[])
    querypath, subjectpath = makefasta(query), makefasta(subject)
    return blastn(querypath, subjectpath, flags)
end

function blastn(query::DNASequence, subject::Vector{DNASequence}, flags=[])
    querypath, subjectpath = makefasta(query), makefasta(subject)
    blastn(querypath, subjectpath, flags)
end

function blastn(query::DNASequence, subject::AbstractString, flags=[]; db::Bool=false)
    querypath = makefasta(query)
    if db
        return blastn(querypath, subject, flags, db=true)
    else
        return blastn(querypath, subject, flags)
    end
end

function blastn(query::Vector{DNASequence}, subject::Vector{DNASequence}, flags=[])
    querypath, subjectpath = makefasta(query), makefasta(subject)
    return blastn(querypath, subjectpath, flags)
end

function blastn(query::Vector{DNASequence}, subject::AbstractString, flags=[]; db::Bool=false)
    querypath = makefasta(query)
    if db
        return blastn(querypath, subject, flags, db=true)
    else
        return blastn(querypath, subject, flags)
    end
end

function blastn(query::AbstractString, subject::Vector{DNASequence}, flags=[])
    subjectpath = makefasta(subject)
    return blastn(query, subjectpath, flags)
end

"""
`blastp(query, subject, flags...)``
Runs blastn on `query` against `subject`.
    Subjects and queries may be file names (as strings), `BioSequence{AminoAcidSequence}` type or
    Array of `BioSequence{AminoAcidSequence}`.
    May include optional `flag`s such as `["-perc_identity", 95,]`. Do not use `-outfmt`.
"""
function blastp(query::AbstractString, subject::AbstractString, flags=[]; db::Bool=false)
    if db
        results = readblastXML(`blastp -query $query -db $subject $flags -outfmt 5`, seqtype = "prot")
    else
        results = readblastXML(`blastp -query $query -subject $subject $flags -outfmt 5`, seqtype = "prot")
    end
    return results
end

function blastp(query::AminoAcidSequence, subject::AminoAcidSequence, flags=[])
    querypath, subjectpath = makefasta(query), makefasta(subject)
    return blastp(querypath, subjectpath, flags)
end

function blastp(query::AminoAcidSequence, subject::Vector{AminoAcidSequence}, flags=[])
    querypath, subjectpath = makefasta(query), makefasta(subject)
    return blastp(querypath, subjectpath, flags)
end

function blastp(query::AminoAcidSequence, subject::AbstractString, flags=[]; db::Bool=false)
    querypath = makefasta(query)
    if db
        return blastp(querypath, subject, flags, db=true)
    else
        return blastp(querypath, subject, flags)
    end
end

function blastp(query::Vector{AminoAcidSequence}, subject::Vector{AminoAcidSequence}, flags=[])
    querypath, subjectpath = makefasta(query), makefasta(subject)
    return blastp(querypath, subjectpath, flags)
end

function blastp(query::Vector{AminoAcidSequence}, subject::AbstractString, flags=[]; db::Bool=false)
    querypath = makefasta(query)
    if db
        return blastp(querypath, subject, flags, db=true)
    else
        return blastp(querypath, subject, flags)
    end
end

function blastp(query::AbstractString, subject::Vector{AminoAcidSequence}, flags=[])
    subjectpath = makefasta(subject)
    return blastp(query, subjectpath, flags)
end

# Create temporary fasta-formated file for blasting.
function makefasta(sequence::BioSequence)
    path, io = mktemp()
    write(io, ">$path\n$(convert(String, sequence))\n")
    close(io)
    return path
end

# Create temporary multi fasta-formated file for blasting.
function makefasta(sequences::Vector{T}) where T <: BioSequence
    path, io = mktemp()
    counter = 1
    for sequence in sequences
        write(io, ">$path$counter\n$(convert(String, sequence))\n")
        counter += 1
    end
    close(io)
    return path
end


dev/search_index.js
var documenterSearchIndex = {"docs": [

{
    "location": "#",
    "page": "Home",
    "title": "Home",
    "category": "page",
    "text": ""
},

{
    "location": "#img-src\"./sticker.svg\"-width\"30%\"-align\"right\"-/-BioTools-1",
    "page": "Home",
    "title": "<img src=\"./sticker.svg\" width=\"30%\" align=\"right\" /> BioTools",
    "category": "section",
    "text": "(Image: Latest Release)  (Image: MIT license)  (Image: Stable documentation)  (Image: Latest documentation) (Image: Lifecycle) (Image: Chat on Discord)"
},

{
    "location": "#Description-1",
    "page": "Home",
    "title": "Description",
    "category": "section",
    "text": "BioTools provides interfaces to common external biological tools from julia scripts and programs."
},

{
    "location": "#Installation-1",
    "page": "Home",
    "title": "Installation",
    "category": "section",
    "text": "Install BioTools from the Julia REPL:using Pkg\nadd(\"BioTools\")\n# Pkg.add(\"BioTools\") for julia v0.6-If you are interested in the cutting edge of the development, please check out the master branch to try new features before release."
},

{
    "location": "#Testing-1",
    "page": "Home",
    "title": "Testing",
    "category": "section",
    "text": "BioTools is tested against Julia 0.7 and current 1.X on Linux, OS X, and Windows.PackageEvaluator Latest Build Status\n(Image: ) (Image: ) (Image: ) (Image: ) (Image: )"
},

{
    "location": "#Contributing-1",
    "page": "Home",
    "title": "Contributing",
    "category": "section",
    "text": "We appreciate contributions from users including reporting bugs, fixing issues, improving performance and adding new features.Take a look at the CONTRIBUTING file provided with every BioJulia package package for detailed contributor and maintainer guidelines."
},

{
    "location": "#Financial-contributions-1",
    "page": "Home",
    "title": "Financial contributions",
    "category": "section",
    "text": "We also welcome financial contributions in full transparency on our open collective. Anyone can file an expense. If the expense makes sense for the development of the community, it will be \"merged\" in the ledger of our open collective by the core contributors and the person who filed the expense will be reimbursed."
},

{
    "location": "#Backers-and-Sponsors-1",
    "page": "Home",
    "title": "Backers & Sponsors",
    "category": "section",
    "text": "Thank you to all our backers and sponsors!Love our work and community? Become a backer.(Image: backers)Does your company use BioJulia? Help keep BioJulia feature rich and healthy by sponsoring the project Your logo will show up here with a link to your website.(Image: ) (Image: ) (Image: ) (Image: ) (Image: ) (Image: ) (Image: ) (Image: ) (Image: ) (Image: )"
},

{
    "location": "#Questions?-1",
    "page": "Home",
    "title": "Questions?",
    "category": "section",
    "text": "If you have a question about contributing or using BioJulia software, come on over and chat to us on Discord, or you can try the Bio category of the Julia discourse site."
},

{
    "location": "blast/#",
    "page": "BLAST",
    "title": "BLAST",
    "category": "page",
    "text": "CurrentModule = BioTools.BLAST"
},

{
    "location": "blast/#The-BioTools-BLAST-wrapper-1",
    "page": "BLAST",
    "title": "The BioTools BLAST wrapper",
    "category": "section",
    "text": "The BioTools.BLAST module is a wrapper for the command line interface of BLAST+ from NCBI. It requires that you have BLAST+  installed and accessible in your PATH (eg. you should be able to execute $ blastn -h from the command line)."
},

{
    "location": "blast/#The-Basics-1",
    "page": "BLAST",
    "title": "The Basics",
    "category": "section",
    "text": "This module allows you to run protein and nucleotide BLAST (blastp and blastn respectively) within julia and to parse BLAST results into Bio.jl types.using BioSequences,\n      BioTools.BLAST\n\nseq1 = dna\"\"\"\nCGGACCAGACGGACACAGGGAGAAGCTAGTTTCTTTCATGTGATTGANAT\nNATGACTCTACTCCTAAAAGGGAAAAANCAATATCCTTGTTTACAGAAGA\nGAAACAAACAAGCCCCACTCAGCTCAGTCACAGGAGAGAN\n\"\"\"\n\nseq2 = dna\"\"\"\nCGGAGCCAGCGAGCATATGCTGCATGAGGACCTTTCTATCTTACATTATG\nGCTGGGAATCTTACTCTTTCATCTGATACCTTGTTCAGATTTCAAAATAG\nTTGTAGCCTTATCCTGGTTTTACAGATGTGAAACTTTCAA\n\"\"\"\n\nblastn(seq1, seq2)These functions return a Vector{BLASTResult}. Each element is a hit which includes the sequence of the hit, an AlignedSequence using the original query as a reference and some additional information (expect vaue, bitscore) for the hit.struct BLASTResult\n    bitscore::Float64\n    expect::Float64\n    queryname::String\n    hitname::String\n    hit::BioSequence\n    alignment::AlignedSequence\nendIf you\'ve already run a blast analysis or have downloaded blast results in XML format from NCBI you can also pass an XML string to readblastXML() in order to obtain an array of BLASTResults.results = readall(open(\"blast_results.xml\"))\n# need to use `readstring` instead of `readall` for v0.5\nreadblastXML(results)When parsing protein blast results, you must include the argument  seqtype=\"prot\", eg. readblastXML(\"results, seqtype=\"prot\")."
},

{
    "location": "blast/#Options-for-blastn-and-blastp-1",
    "page": "BLAST",
    "title": "Options for blastn and blastp",
    "category": "section",
    "text": "Both of the basic BLAST+ commands can accept a single BioSequence, a Vector{BioSequence} or a sting representing a file path to a fasta formatted file as arguments for both query and subject.blastn([seq1, seq2], [seq2, seq3])\n\nblastp(aaseq, \"path/to/sequences.fasta\")If you have a local blast database (eg through the use of $ makeblastdb), you can use this database as the subjectblastn(seq1, \"path/to/blast_db\", db=true)If you want to modify the search using additional options (eg. return only results with greater than 90% identity), you may pass a Vector of flags (see here for valid arguments - do not use flags that will alter file handling such as -outfmt)blastn(seq1, seq2, [\"-perc_identity\", 90, \"-evalue\", \"9.0\"])"
},

]}
